src/Compiler/test/Microsoft.NET.Sdk.Razor.SourceGenerators.Tests/RazorSourceGeneratorTestsBase.cs

```diff
@@ -154,22 +168,50 @@ protected static async Task<string> RenderRazorPageAsync(Compilation compilation
             writer,
             new HtmlHelperOptions());
 
-        page.ViewContext = viewContext;
-        page.HtmlEncoder = HtmlEncoder.Default;
+        // Find `_ViewStart.cshtml`s.
+        var viewStarts = GetViewStartNames(name)
+            .Select(n => assembly.GetType($"{generatedNamespace}.{n}"))
+            .Where(t => t is not null)
+            .Select(t => (IRazorPage)Activator.CreateInstance(t!)!)
+            .ToImmutableArray();
 
         // Render the page.
-        await page.ExecuteAsync();
+        var view = ActivatorUtilities.CreateInstance<RazorView>(app.Services,
+            /* IReadOnlyList<IRazorPage> viewStartPages */ viewStarts,
+            /* IRazorPage razorPage */ page);
+        await view.RenderAsync(viewContext);
 
         assemblyLoadContext.Unload();
 
         return writer.ToString();
+
+        // Inspired by Microsoft.AspNetCore.Mvc.Razor.RazorFileHierarchy.GetViewStartPaths.
+        static IEnumerable<string> GetViewStartNames(string name)
+        {
+            var builder = new StringBuilder(name);
+            var maxIterations = 255;
```

Comment:
const, and a `currentIteration` counter for the current variable? I would also probably prefer a `for` loop that doesn't rely on a `--` inline.

Reply:
Sure, that looks better, thanks. (I was just trying to be consistent with the original implementation I copied.)